feat(atc): support var_sources in pipeline.

evanchaoli opened the pull request:
Resolve #4579 

This is for initial review from design perspective. 

Highlights of the changes:

1. Add "credential_managers" in pipeline YAML, `atc/config.go`
2. Valid credential manager's configure in `atc/validate.go`. Also to solve cyclic imports, moved `validate.go` to `atc/configvalidate/validate.go`.
3. Add a new field `credential_managers` in table `pipelines`. `fly set-pipeline` will store pipeline credential managers to this new field.
4. Add a new field `credential_managers` in table `builds`. When create new build, get the value from `pipeline`.
5. In `atc/engine/builder.go`, create pipeline secrets variables, put them and global secrets into a `MultiVars`, so that build steps can interpolate secrets from pipeline credential managers as well as the global credential manager.
6. Add a new function to interface `creds.Manager`. When a build finishes, call this method to cleanup pipeline credential managers, basically, stop the `reauther` go routine.

Note: Tests have not been updated yet.


# Contributor Checklist
> Are the following items included as part of this PR? Please delete checkbox items that don't apply.
- [ ] Unit tests
- [ ] Integration tests (if applicable)
- [ ] Updated documentation (located at https://github.com/concourse/docs)
- [ ] Updated release notes (located at https://github.com/concourse/concourse/tree/master/release-notes)


# Reviewer Checklist
> This section is intended for the core maintainers only, to track review progress. Please do not
> fill out this section.
- [ ] Code reviewed
- [ ] Tests reviewed
- [ ] Documentation reviewed
- [ ] Release notes reviewed
- [ ] PR acceptance performed

vito commented:
Hey thanks for picking this up! :slightly_smiling_face: Hoping to get to this soon and submit some feedback. I've got a few loose-ends to tidy up in in this area, so I'm going to address them beforehand just to make sure my head's in the right place with regard to the long-term vision:

1. https://github.com/concourse/rfcs/pull/21 which is likely to be superceded by...
1. https://github.com/concourse/rfcs/pull/32 which introduces project-level credential manager config (as `vars_sources:`)

Once I have this all cleaned up I'll come back and cross-reference your proposal with mine and make sure there's common ground and a clear path forward to the long-term roadmap. :slightly_smiling_face:

My tentative thoughts are that it's fine to introduce them at the pipeline-level first since projects will take a while to get here, and starting with pipeline-level creds will at least get us away from a single global credential manager.

I'd also be excited to use this as a cheaper way of testing credential management. Perhaps in combination with #4502 we can even add a `testflight` test and/or start to test Vault and other integrations more cheaply (now that they can be tested independent of the cluster's configuration).

Thanks again! I'll try to get to this soon.

evanchaoli commented:
I uses two vaults in my test:
* `http://<local-ip>:7200` is configured to `concourse web`
* `http://<local-ip>:8200` is configured in pipeline yaml

Pipeline:

    credential_managers:
    - name: dev-vault
      type: vault
      config:
        url: http://192.168.205.77:8200
        path_prefix: concourse
        client_token: ((my-vault-token))
    
    jobs:
    - name: test
      public: true
      plan:
      - task: task1
        config:
          platform: linux
          image_resource:
            type: registry-image
            source: {repository: busybox}
          run:
            path: sh
            args:
             - -exc
             - |
           echo The name is ((test.pass))
 
The pipeline Vault's config takes a variable `((my-vaule-token))`, that will be interpolated from atc's vault, and the task takes a variable `((test.pass))`, that will be from the pipeline vault.

The screen shots of my test result is simple, but it proves the change has been working:

<img width="1059" alt="test-multiple-vaults" src="https://user-images.githubusercontent.com/18585861/66952522-a54e6780-f011-11e9-8f87-a74e0ac7ba54.png">

vito commented:
> The pipeline Vault's config takes a variable `((my-vaule-token))`, that will be interpolated from atc's vault, and the task takes a variable `((test.pass))`, that will be from the pipeline vault.

Hah, that's pretty cool. I was wondering about this part because these pipeline-level vars would otherwise need to be directly embedded (scary even with encryption). So with this it's sort of like having a hierarchy of credential management. :thinking: An interesting pattern. Up until now I've been hoping to remove the global credential manager, but I guess this would be one reason to keep it around.

pivotal-jwinters commented on atc/db/migration/migrations/1570709099_add_credential_managers_to_pipelines.up.sql:
```diff
@@ -0,0 +1,4 @@
+BEGIN;
+  ALTER TABLE pipelines ADD COLUMN credential_managers json;
```
Can we actually make credential_managers its own table? We could probably model them very similarly to resources.

evanchaoli replied:
@pivotal-jwinters Thanks for the comment. I considered that at the beginning, and I decided to begin with a simple way as this doesn't impact big picture design of this feature. So I'd defer this change till other design parts get clear.

evanchaoli commented on atc/api/configserver/get.go:
```diff
@@ -64,10 +64,11 @@ func (s *Server) GetConfig(w http.ResponseWriter, r *http.Request) {
-		ResourceTypes: resourceTypes.Configs(),
-		Jobs:          jobs.Configs(),
+		Groups:             pipeline.Groups(),
+		CredentialManagers: pipeline.CredentialManagers(),
```
No problem. I'll rename it in next push.

evanchaoli commented on atc/configvalidate/validate.go:
```diff
@@ -679,3 +683,42 @@ func compositeErr(errorMessages []string) error {
+			return fmt.Errorf("unknown credential manager type: %s", cm.Type)
+		}
+
+		// TODO: this check should eventually be removed once all credential managers
```
Yes, `.Close`, `.NewInstance` and `MarshalJSON`, but those are easy part. I think the major thing is testing. As I don't have testing environment for other credential managers, and lack of experience with them, so I'd like ask for help. We can create a "help-wanted" issue for each credential manager, then original authors, or who else familiar with them, may help.

evanchaoli commented on atc/creds/vault/manager.go:
```diff
@@ -84,6 +87,59 @@ func (manager *VaultManager) MarshalJSON() ([]byte, error) {
+	manager.SharedPath = to_string(data["shared_path"])
+
+	manager.TLS.CACert = to_string(data["ca_cert"])
+	manager.TLS.CAPath = to_string(data["ca-path"])
```
Yes, that's a typo. Will fix in next push.

evanchaoli commented on atc/creds/vault/manager.go:
```diff
@@ -84,6 +87,59 @@ func (manager *VaultManager) MarshalJSON() ([]byte, error) {
 	})
 }
 
+func to_string(s interface{}) string {
```
Will fix in next push.

evanchaoli commented on atc/configvalidate/validate.go:
```diff
@@ -679,3 +683,42 @@ func compositeErr(errorMessages []string) error {
+			return err
+		}
+
+		err = json.Unmarshal(b, &manager)
```
Let me investigate a little bit. Fortunately this doesn't affect design of this feature.

pivotal-jwinters replied:
You can have a look at the skymarshal bits [here](https://github.com/concourse/concourse/blob/master/skymarshal/skycmd/flags.go#L77).

It's not pretty, but it does reduce the amount of boilerplate you need to write when adding new auth connectors.

evanchaoli replied:
@pivotal-jwinters and @vito To be honest, I don't clearly the point of `shymarshal`. I check code that Joshua pointed out, looks like the key is to use `github.com/mitchellh/mapstructure`. But with that, we will still need to use `mapstructure`, because: 1) we need to squash struct because `TLS` and `AuthConfig` have second level data, 2) parameter name is not same to struct field name, like "ca-path" vs "CAPath".

For `spf13/viper`, I guess I cannot just use it for `var_sources`, so I cannot use it for this PR.

Did I check the right thing you want to me check?

vito replied:
@evanchaoli Let's just skip over this for now, just do what's easiest and low-ish maintenance and we can re-assess things later on. I think the number of credential managers to support will be lower than the number of auth providers so this hopefully won't be as painful. :)

evanchaoli commented on atc/creds/vault/manager.go:
```diff
@@ -84,6 +87,59 @@ func (manager *VaultManager) MarshalJSON() ([]byte, error) {
+	return s.(time.Duration)
+}
+
+func (manager *VaultManager) UnmarshalJSON(text []byte) error {
```
I considered this. The reason why I decided to write `UnmarshalJSON` is that I want to Vault config to be consistent with CLI options. For example, in CLI, there is a option `--vault-client-token`, if I use json tag, then in yaml it should be:

    config: 
       auth:
          client_token: xxx

I want one level config:

    config:
       client_token: xxx

so that, user just needs to check `concourse web --help`, then he knows how to config in yaml.

vito replied:
I was hoping there was a `json:",inline"` of sorts but it doesn't seem to exist, so :+1: for now. I think this kind of goes hand-in-hand with my other comment so we can skip over this too.

evanchaoli commented on atc/creds/vault/manager.go:
```diff
@@ -84,6 +87,59 @@ func (manager *VaultManager) MarshalJSON() ([]byte, error) {
+	manager.Auth.BackendMaxTTL = to_duration(data["auth_max_ttl"], 0)
+	manager.Auth.RetryMax = to_duration(data["auth_retry_max"], 5*time.Minute)
+	manager.Auth.RetryInitial = to_duration(data["auth_retry_initial"], 1*time.Second)
+	if data["auth-param"] == nil {
```
Will fix in next push.

evanchaoli commented on atc/creds/vault/manager.go:
```diff
@@ -94,6 +150,10 @@ func (manager VaultManager) Validate() error {
 	}
 
+	if manager.PathPrefix == "" {
+		return fmt.Errorf("path prefix must be a non-empty string")
```
For the global vault, there is default `/concourse`, but for pipeline vault, user may forget to set path prefix, that's reason why I add this check.

vito replied:
Makes sense to not have a default path prefix for pipelines since the need for it globally is sort of from the nature of the 'global credential manager' flow. +1 to the validation then.

evanchaoli commented on atc/creds/vault/manager.go:
```diff
@@ -120,7 +180,14 @@ func (manager VaultManager) Health() (*creds.HealthResponse, error) {
-	ra := NewReAuther(manager.Client, manager.Auth.BackendMaxTTL, manager.Auth.RetryInitial, manager.Auth.RetryMax)
-	return NewVaultFactory(manager.Client, ra.LoggedIn(), manager.PathPrefix, manager.SharedPath), nil
+func (manager *VaultManager) NewSecretsFactory(logger lager.Logger) (creds.SecretsFactory, error) {
+	if manager.SecretFactory == nil {
```
If multiple builds are triggered at the same time, it would be a race condition. I'll add a lock in next push.

evanchaoli replied:
With my latest change, this will be called by `VarSourcePool` who has a lock, thus no need to lock here.

evanchaoli replied:
With my latest change,  this function will be called from `VarSourcePool`. The caller will lock, thus no need to lock here.

evanchaoli commented on atc/db/build.go:
```diff
@@ -184,6 +208,33 @@ func (b *build) IsRunning() bool      { return !b.completed }
+func (b *build) CredentialManagers() (atc.CredentialManagerConfigs, error) {
+	var credentialManagers sql.NullString
+
+	err := psql.Select("credential_managers").
```
No problem, will add encryption in next push.

evanchaoli commented on atc/db/build.go:
```diff
@@ -1185,6 +1259,27 @@ func (b *build) saveEvent(tx Tx, event atc.Event) error {
 func createBuild(tx Tx, build *build, vals map[string]interface{}) error {
 	var buildID int
+
+	// See if the pipeline has credential managers configured, ff yes, add to build.
```
In some case, a pipeline credential manager may change. After the change, if there are still pending builds, then old builds should still use old credential manager. That's reason why I copy the info from `pipelines` to `builds`.

vito replied:
:thinking: Do you feel that this will be a common occurrence worth accounting for? I feel like I wouldn't be too shocked if, as a user, I updated the credential manager config and my older in-flight builds adopted the new changes. For example, maybe I just rotated my Vault credentials. Do you expect something different from your workflow?

evanchaoli replied:
Actually I don't have strong opinion on this. Updating Vault configure should not be something happening now and again. I can remove `var_sources` from `builds`.

evanchaoli replied:
Removed `var_sources` from `builds`.

evanchaoli commented on atc/engine/builder/builder.go:
```diff
@@ -43,23 +46,35 @@ func NewStepBuilder(
+	if builder.pipelineCredMgrs[build.ID()] != nil {
+		logger.Debug("close-pipeline-credential-managers")
+		for _, cm := range builder.pipelineCredMgrs[build.ID()] {
+			cm.Close(logger)
```
I think pipeline is a static thing, a pipeline may not have any build for long time, it makes not much sense to maintain credential managers for no-build pipeline.

A credential manager just launches a go-routine, a build already has a lot of go-routines, so one more seems not much matter.

vito replied:
Spending one goroutine is fine, but there's probably a heavier cost than just a single goroutine, since these goroutines create connections and that may involve doing a TLS handshake repeatedly if it's a fresh client every time.

Users have observed that constantly doing TLS handshakes can actually be very costly when talking to Postgres; increasing the idle connection pool size significantly reduced CPU usage on the `web` nodes because they didn't have to do the TLS handshake as often.

In looking at our Vault code, we use Vault with its default 'pooled' client, which will probably help mitigate the TLS handshake cost by re-using connections, but it actually warns that it shouldn't be used for 'transient clients':

https://github.com/hashicorp/vault/blob/ad3f31d7954a1c46b2c6e01f0c72a033ffd1117a/api/client.go#L142

https://github.com/hashicorp/vault/blob/ad3f31d7954a1c46b2c6e01f0c72a033ffd1117a/vendor/github.com/hashicorp/go-cleanhttp/cleanhttp.go#L49-L52

Now, this *might* not be a big deal because we'll likely be re-using the pooled client for a stable-ish set of hosts over time, but it's still a little worrying.

We don't have to solve this right off the bat, since it might be OK, but I could see it behaving poorly at medium-to-large scale. Maybe we just need to test it out at scale and see where it bends and/or breaks. :thinking:

vito replied:
@evanchaoli Actually, won't we have to use these pipeline credential managers for resource checking, too? At that point we would basically always have an active client. Is that not implemented in the PR yet?

evanchaoli replied:
> @evanchaoli Actually, won't we have to use these pipeline credential managers for resource checking, too? At that point we would basically always have an active client. Is that not implemented in the PR yet?

It's not implemented yet in this PR. Actually this is one thing I was not sure, I added the question at https://github.com/concourse/concourse/pull/4600/files#diff-e6d6a569fec760fc128319f5e4eda027R161. `check` step only interpolates variables for resource configure, right?

vito replied:
Yep!

evanchaoli replied:
@vito I checked `check` related code, now I agree pipeline credential managers should be maintained at pipeline level. Given multiple pipelines might use same credential manager, I'm thinking to create a global credential manager pool (map), using interpolated `var-source-config` as key, so that we can avoid duplicate credential manager instances. One question, what's plan of making `lidar` as default? In other words, do we need to support this feature for  non-lidar?

vito replied:
@evanchaoli We plan to make `lidar` the default ASAP (based on user feedback as we get more mileage on it), at least by the time we ship v6.0, so I'm OK with not supporting non-lidar (aka radar).

A global credential manager pool makes sense to me! Would you be trying @pivotal-jwinters's suggestion of a `credential_managers` table?

FWIW you'll probably want to key on a hash of the config, similar to how we key resource configs.

There's probably more to figure out than I'm able to cover at this point, but we can figure it out as we go. :)

evanchaoli replied:
I have added a `VarSourcePool` in `creds/pool.go`, it maintains a map of credential managers, key is config of managers. I have made both lidar and non-lidar work now.

evanchaoli replied:
Yes, I have done keying on the config in the commit I just pushed. Also I have made both lidar and non-lidar `check`s work.

evanchaoli commented:
@vito I have just pushed a new version, and resolved some review comments.

evanchaoli commented:
@vito I just pushed a new commit. Highlights of this new commit:

* Add `VarSourcePool` in `atc/creds/pool.go`. It maintains a map of credential managers by config. So that if multiple pipelines configure same Vault, then only one Vault instance will be created.
* `VarSourcePool` is a `runner`. It periodically check `lastUseTime` of all managers. If a manager is not used for certain time (I hard code 5 minutes now, we can make it configurable), it will be closed and deleted from the map.
* Add `Variables()` to struct `Pipeline`, so that `var_sources` is handled at pipeline level.
* Make `check` of both lidar and non-lidar work now.
* Add interface `PipelineRef` in `atc/db/pipeline_ref.go` in order to avoid duplicate code of pipeline reference in many structures.

In this version, I might not clean up code perfectly. The code may contain some commented lines, "EVAN" in logs, etc. Please ignore those nits. I'll clean up and squash later.

vito commented on atc/engine/engine.go:
```diff
@@ -175,12 +180,20 @@ func (b *engineBuild) Run(logger lager.Logger) {
+		// Fails the build if BuildStep returned error. Because some unrecoverable error,
+		// like pipeline credential manager is wrong, will cause a build to never start
+		// to run.
+		builder.NewBuildStepDelegate(
```
This kinda sticks out like a sore thumb. :thinking: How likely are these errors to occur?

evanchaoli replied:
Without this feature, `BuildStep` seems to rarely fail though there are a few clauses returning error. But with this feature, if, like Vault config in pipeline yaml is wrong (meaning invalid vault URL, not something can be blocked by `Validate`), then `BuildStep` will fail, this kind of error could be very possible. I think, in any case, as long as `BuildStep` returns error, then error should be handled.

vito replied:
I just meant that the constructing of the delegate in-place and calling a method on it is kind of weird looking and probably hard to cover with a test. I agree though, it would be nice to surface to the user. If you could add a unit test for this, which I suspect would involve some refactoring, I'm good with it. :+1:

Maybe the error logging could be moved into the builder somehow since it already has a delegate factory injected? :man_shrugging: Might be some other way to clean it up too.

vito commented on atc/radar/resource_scanner.go:
```diff
@@ -111,6 +111,14 @@ func (scanner *resourceScanner) scan(logger lager.Logger, resourceID int, fromVe
 
+	found, err = scanner.dbPipeline.Reload()
+	if !found {
+		return 0, fmt.Errorf("Pipeline %s(%d) not found", scanner.dbPipeline.Name(), scanner.dbPipeline.ID())
```
```suggestion
		return 0, fmt.Errorf("pipeline %s(%d) not found", scanner.dbPipeline.Name(), scanner.dbPipeline.ID())
```

vito commented on atc/radar/resource_scanner.go:
```diff
@@ -111,6 +111,14 @@ func (scanner *resourceScanner) scan(logger lager.Logger, resourceID int, fromVe
+		return 0, fmt.Errorf("Pipeline %s(%d) not found", scanner.dbPipeline.Name(), scanner.dbPipeline.ID())
+	}
+	if err != nil {
+		return 0, fmt.Errorf("Failed to reload pipeline %s(%d)", scanner.dbPipeline.Name(), scanner.dbPipeline.ID())
```
```suggestion
		return 0, fmt.Errorf("failed to reload pipeline %s(%d)", scanner.dbPipeline.Name(), scanner.dbPipeline.ID())
```

vito commented on atc/radar/resource_type_scanner.go:
```diff
@@ -86,6 +87,14 @@ func (scanner *resourceTypeScanner) scan(logger lager.Logger, resourceTypeID int
 
+	found, err = scanner.dbPipeline.Reload()
+	if !found {
+		return 0, fmt.Errorf("Pipeline %s(%d) not found", scanner.dbPipeline.Name(), scanner.dbPipeline.ID())
```
```suggestion
		return 0, fmt.Errorf("pipeline %s(%d) not found", scanner.dbPipeline.Name(), scanner.dbPipeline.ID())
```

vito commented on atc/radar/resource_type_scanner.go:
```diff
@@ -86,6 +87,14 @@ func (scanner *resourceTypeScanner) scan(logger lager.Logger, resourceTypeID int
+		return 0, fmt.Errorf("Pipeline %s(%d) not found", scanner.dbPipeline.Name(), scanner.dbPipeline.ID())
+	}
+	if err != nil {
+		return 0, fmt.Errorf("Failed to reload pipeline %s(%d)", scanner.dbPipeline.Name(), scanner.dbPipeline.ID())
```
```suggestion
		return 0, fmt.Errorf("failed to reload pipeline %s(%d)", scanner.dbPipeline.Name(), scanner.dbPipeline.ID())
```

vito commented on atc/utils/stringencrypt.go:
```diff
@@ -0,0 +1,95 @@
+package utils
```
Guessing this whole thing can be removed once everything's switched over to the preexisting encryption stack. :)

vito commented on atc/db/build.go:
```diff
@@ -55,12 +79,12 @@ var minMaxIdQuery = psql.Select("COALESCE(MAX(b.id), 0)", "COALESCE(MIN(b.id), 0
 //go:generate counterfeiter . Build
 
 type Build interface {
+	PipelineRef
```
This seems like a cool idea, but it looks like it's not checked in. :)

vito commented on atc/atccmd/command.go:
```diff
@@ -873,6 +873,14 @@ func (cmd *RunCommand) constructBackendMembers(
 		)},
+		{Name: "var-sources-watcher", Runner: lockrunner.NewRunner(
+			logger.Session("var-sources-watcher"),
+			creds.VarSourcePoolInstance(),
```
This type doesn't seem to be checked in either. :)

evanchaoli commented:
@vito Really sorry for missing to check in the two new files, I always forget `git add` new files. (: I just quickly cleaned up some nit comments, and added the two missing files. I will address rest of comments soon.

pivotal-jamie-klassen commented:
I was just perusing old PRs and saw one that reminded me of this work. I might have misunderstood the core concepts here, but if this gets merged, might it unblock #3081?

evanchaoli commented:
@vito I just pushed again and addressed all of comments including data encryption and `NewInstance(map[string]interface{}` issues. I should have pushed last night, but unfortunately the hotel's wifi seems to block github.com, thus I came to office this morning and got the code pushed.

evanchaoli commented:
> I was just perusing old PRs and saw one that reminded me of this work. I might have misunderstood the core concepts here, but if this gets merged, might it unblock #3081?

I think this PR will unblock #3081. Once this PR is merged, user may define credential in pipeline, but #3081 still needs to add `key-vault` as a credential manager.

vito commented:
@evanchaoli Taking another look at this today! One quick request: would you mind making subsequent changes as new commits instead of force-pushing? It'll just make it a bit easier for me to tell what's new when re-reviewing. Thanks!

evanchaoli commented on atc/creds/dummy/manager_factory.go:
```diff
@@ -27,3 +28,12 @@ func (factory *managerFactory) AddConfig(group *flags.Group) creds.Manager {
 	return manager
 }
+
+func (factory *managerFactory) NewInstance(config interface{}) (creds.Manager, error) {
```
I agree!

evanchaoli commented on atc/creds/pool.go:
```diff
@@ -0,0 +1,110 @@
+			lastUseTime:    time.Now(),
+		}
+	} else {
+		logger.Debug("found-existing-credential-manager")
```
Good catch!

evanchaoli commented on atc/creds/pool.go:
```diff
@@ -0,0 +1,110 @@
+}
+
+func VarSourcePoolInstance() *varSourcePool {
+	return &pool
```
@vito I just created a quick test program and verified:

```go
type someData struct {
	i int
}

var data = someData{i: 1}

func dataObj() *someData {
	return &data
}

func TestData() {
	dataObj().i = 5
	fmt.Printf("i=%d\n", data.i)
	dataObj().i = 8
	fmt.Printf("i=%d\n", data.i)
}
```
Output is:
```
i=5
i=8
```

So it won't copy `data.i`.

Any to make you comfortable, I can make `pool` as pointer in the first place. :)

evanchaoli commented on atc/creds/dummy/manager_factory.go:
```diff
@@ -27,3 +28,12 @@ func (factory *managerFactory) AddConfig(group *flags.Group) creds.Manager {
+
+func (factory *managerFactory) NewInstance(config interface{}) (creds.Manager, error) {
+	if _, ok := config.(map[string]interface{}); !ok {
+		return nil, fmt.Errorf("invalid vault config format")
```
Bad thing from copy-paste.

evanchaoli commented on atc/db/migration/migrations/1570709099_add_varsources_to_pipelines.up.sql:
```diff
@@ -0,0 +1,4 @@
+BEGIN;
+  ALTER TABLE pipelines ADD COLUMN var_sources text,
+                        ADD COLUMN nonce text;
```
I actually thought over that when I added the `nonce` column. It should be named `foo_nonce`, but given there is only one column to encrypt, I though to keep consist to other tables, so that at least which looks better for now. As you have pointed the concern, let me rename the column, so that reduce one worm when you open the can.

evanchaoli commented on atc/creds/pool.go:
```diff
@@ -0,0 +1,110 @@
+	return pool.pool[key].NewSecrets(), nil
+}
+
+func (pool *varSourcePool) Run(ctx context.Context) error {
```
I guess renaming the method will bring certain efforts, as the `interface Task` is defined in `atc/lockrunner/runner.go`, which is not specific to GC collector.

From my side, what I can do is to rename `varSourcePool.Run()` to something like `varSourcePool.Clean()`, and add a `varSourcesCollector` under `atc\gc`. But I don't really like that, because it would just introduce some unnecessary duplicate code.

@vito WDYT?

evanchaoli commented on atc/db/pipeline_ref.go:
```diff
@@ -0,0 +1,64 @@
+package db
```
Thanks :)

evanchaoli commented on atc/creds/credhub/credhub.go:
```diff
@@ -18,12 +18,15 @@ type CredHubAtc struct {
 		lookupPaths = append(lookupPaths, creds.NewSecretLookupWithPrefix(path.Join(c.prefix, teamName, pipelineName)+"/"))
 	}
 	lookupPaths = append(lookupPaths, creds.NewSecretLookupWithPrefix(path.Join(c.prefix, teamName)+"/"))
+	if allowRootPath {
```
@vito WRT [concourse/rfcs#21 (comment)](https://github.com/concourse/rfcs/pull/21#discussion_r269687922), I think that makes sense for team level credential manager, because that would be configured by concourse admin; but doesn't make that much sense at pipeline level, because it adds overhead for pipeline author, and hurts mobility of pipelines, consider if a pipeline yml is shared by multiple teams or multiple pipelines, or migrate a pipeline from one concourse instance to the other.

For us, I'd keep current `NewSecretLookupPaths` pattern. As Concourse doesn't support sub-team, a big BU may have multiple teams on our Concourse cluster, but the BU has a single Vault, so each team will still put their secrets under their own Vault path. At Vault side, permission management is the other headache, who can access which path, which is complex. Also given the nature of Security, migrating secrets of Vault, from one path to other path, from one Vault instance to other instance is not easy, not much tool can help. So I'd be very careful for any Vault related changes. As @ralekseenkov mentioned, it should not be any breaking changes.

evanchaoli commented:
> I got the dummy credential manager working and wrote a `testflight` test using it - here's a PR to your fork: [evanchaoli#1](https://github.com/evanchaoli/concourse/pull/1) - theoretically if that's merged it'll propagate to this PR too.

@vito I have never done this cool stuff before. I should have merged your PR first, then address review comment. Unfortunately I did in reverse order, so that generated a lot of conflicts when merged your PR. Anyway, this is a good lesson learned.

Looks like there are not much architecture level comment, I'm going to update unit tests.

vito commented on atc/creds/pool.go:
```diff
@@ -61,11 +63,9 @@ func (pool *varSourcePool) FindOrCreate(logger lager.Logger, config map[string]i
 		pool.pool[key] = &inPoolManager{
 			manager:        manager,
 			secretsFactory: secretsFactory,
-			lastUseTime:    time.Now(),
```
Why was this removed now? :thinking:

evanchaoli replied:
Because `lastUseTime` is updated in `inPoolManager.NewSecrets()`.

vito replied:
Ah got it, thanks!

vito commented on atc/creds/pool.go:
```diff
@@ -0,0 +1,107 @@
+	pool.lock.Lock()
+	defer pool.lock.Unlock()
+
+	logger.Debug("before collect", lager.Data{"pool-size": len(pool.pool)})
```
this should be done using a session like so:

```go
logger = logger.Session("collect", lager.Data{"pool-size": len(pool.pool)})

logger.Debug("start")
defer logger.Debug("done")
```

also, log messages should be hyphen-delimited, though in the above example that becomes a moot point. it's kind of weird but it's the [`lager`](https://github.com/cloudfoundry/lager) way.

evanchaoli replied:
@vito We cannot put `len(pool.pool)` to session, because pool size might change after collecting. The reason why I add pool-size in logs is to track pool size before and after collect.

Also, when `Collect` is called, the runner has started a session and only passes session logger to `Collect`, so I don't think we should start a session again in `Collect`.

I have changed log actions to hyphen-delimited.

vito commented on atc/creds/pool_test.go:
```diff
@@ -0,0 +1,82 @@
+				Expect(found).To(BeFalse())
+			})
+		})
+	})
```
can tests for `Collect` be added? or is it covered elsewhere?

evanchaoli replied:
Nice comment!!! :) I tried to be "lazy" by using the singleton pool. To add test case of `Collect`, I have to move away from the singleton pool, instead create `varSourcePool` in cmd and pass to everywhere. Anyway, I agree this makes the code better. :)

vito commented on atc/creds/vault/reauther_test.go:
```diff
@@ -46,7 +47,8 @@ func testWithoutVaultErrors(t *testing.T) {
 		Delay:        1 * time.Second,
 	}
-	ra := NewReAuther(ma, 10*time.Second, 1*time.Second, 64*time.Second)
+	logger := lager.NewLogger("vault-test")
```
you can use `lagertest.NewTestLogger` here, which will log to `GinkgoWriter`

vito commented on atc/creds/vault/vault_test.go:
```diff
@@ -147,5 +147,30 @@ var _ = Describe("Vault", func() {
+				variables = creds.NewVariables(v, "team", "pipeline", true)
+			})
+
+			It("should not get secret from root", func() {
```
should probably have two examples: one showing it can and one showing it can't. this one seems to say it can't but then demonstrates that it does (because it's given `true` in the `JustBeforeEach`)

vito commented on atc/creds/vault/vault_test.go:
```diff
@@ -147,5 +147,30 @@ var _ = Describe("Vault", func() {
 		})
+
+		Context("allowRootPath", func() {
+			JustBeforeEach(func() {
```
nit: `JustBeforeEach` should only really be used for performing the effect under test, and there should only be one per context; otherwise just use `BeforeEach` for setup

evanchaoli replied:
@vito I used `JustBeforeEach` here because the original test code initially used `JustBeforeEach` at line#39. OK, let me change all to `BeforeEach`.

vito commented on atc/db/migration/migrations/1570709099_add_varsources_to_pipelines.up.sql:
```diff
@@ -0,0 +1,4 @@
+BEGIN;
+  ALTER TABLE pipelines ADD COLUMN var_sources text,
+                        ADD COLUMN var_sources_nonce text;
```
Sorry, I shouldn't have suggested this yet; currently the column has to be called `nonce` because our encryption code hardcodes it here:

https://github.com/concourse/concourse/blob/23c12778b8c424f9ed9f009154461b0bd321c755/atc/db/open.go#L136-L139

Also this column will need to be added to the list of encrypted columns here:

https://github.com/concourse/concourse/blob/23c12778b8c424f9ed9f009154461b0bd321c755/atc/db/open.go#L123-L131

This code is for rotating encryption keys on startup.

I'm OK with just calling it `nonce` unless you want to also refactor the encryption code to support other column names. (I doubt you're interested in that since you'd also need to migrate all the existing columns.)

vito commented on docker-compose.yml:
```diff
@@ -29,6 +29,11 @@ services:
       CONCOURSE_ADD_LOCAL_USER: test:test,guest:guest
       CONCOURSE_MAIN_TEAM_LOCAL_USER: test
       CONCOURSE_CLUSTER_NAME: dev
+      # FIXME - EVAN: delete VAULT vars.
```
These can probably be removed now so that testflight/watsjs builds can run.

pivotal-jwinters commented on atc/db/migration/migrations/1570709099_add_varsources_to_pipelines.up.sql:
```diff
@@ -0,0 +1,4 @@
+BEGIN;
+  ALTER TABLE pipelines ADD COLUMN var_sources text,
```
Can we also consider my original comment about moving these into their own table. I know it's not a big difference now, but if we ever want to think about de-duplicate credential managers across pipelines (similar to global resources), it will be much easier if these are already split out.

vito replied:
I think they're already de-duped across pipelines via the var sources pool; I'm not really sure if putting them in a separate table buys us anything at that point.

pivotal-jwinters replied:
oh i see now. Each ATC doing their own in memory deduplication and running their own auth loop seems kinda scary to me.

pivotal-jwinters replied:
I'm not too familiar with how the credential managers work, but I'm guessing the vault cred manager already runs 1 auth loop per atc?

vito replied:
Yep - one per ATC today.

In this PR they're "GC'd" from memory after 5 minutes of inactivity, at least.

I think it kind of makes sense since you'd want some form of caching of active clients anyway, so you're not logging in every single time and hammering the credential manager. Most pipelines will be continuously using their credential manager for periodic checking anyway.

The memory usage + active auth loops concern is valid, though it seems kind of tricky to do anything else. If we had a `var_sources` table we could free up the memory use and cache the currently-active credential there instead so we don't have a bunch of auth loops. The downside is then we'd be persisting credentials into the database... which we can encrypt, but I wonder how people would feel about that.

It seems like this `VarSourcePool` object would be a pretty easy place to attempt this sort of thing in the future, which is cool at least!

pivotal-jwinters replied:
I've actually been thinking about what it might look like to run all third party integrations (auth, creds, metrics) as containers on workers. Where each one could be represented by your new `prototype` concept and come with some pre baked actions.

evanchaoli replied:
As a user, I would keep secrets in memory only.

evanchaoli commented:
> What do you think of adding support for other credential managers? Are we planning to just merge this first and add support for the rest later?
> 

@vito I would prefer to merge current code first, because:

* We only use Vault, so I'm not familiar with other credential managers. To support them, it requires both learning curve and environment setup, it would be out of my ability and capacity.
* This PR has been too big, I'd try to avoid to make it bigger.
* This PR has a lot of conflicts to the `set_pipeline` PR #4708, so getting this merged will help move forward that PR.

So I'd suggest merge this PR first, then open a "help-wanted" issue for each credential manager.

Today I have addressed all review comments. In order to add unit test for `pool.Collect`, I had to get rid of the singleton pool instance, thus touched a lot of files to pass the `varSourcePool` to everywhere.

vito commented on atc/radar/resource_scanner.go:
```diff
@@ -155,8 +166,14 @@ func (scanner *resourceScanner) scan(logger lager.Logger, resourceID int, fromVe
 	}
 
+	// Combine pipeline specific var_sources with the global credential manager.
+	varss, err := scanner.dbPipeline.Variables(logger, scanner.variables, scanner.varSourcePool)
```
Ah, I see, this part only has `scanner.variables`, not a secrets manager. Maybe that could be changed. We're going to phase out `radar/` soon-ish anyway. :thinking:

evanchaoli replied:
Changed radar to pass secrets instead of variables.

vito commented on atc/radar/resource_type_scanner.go:
```diff
@@ -116,8 +128,13 @@ func (scanner *resourceTypeScanner) scan(logger lager.Logger, resourceTypeID int
 		return 0, err
 	}
 
+	varss, err := scanner.dbPipeline.Variables(logger, scanner.variables, scanner.varSourcePool)
```
(same here)

vito commented on atc/config.go:
```diff
@@ -40,6 +43,24 @@ func (groups GroupConfigs) Lookup(name string) (GroupConfig, int, bool) {
 }
 
+type VarSourceConfig struct {
+	Name   string      `json:"name"`
```
Actually, is this name used anywhere?

It seems like it would make sense to be able to explicitly reference the var source to fetch from, e.g. `((myvault:foo/bar.baz))`. But I don't think that's been implemented.

I think we should either implement that or remove the field for now.

evanchaoli replied:
`((myvault:foo/bar.baz))` is exactly what one of our users asked to me when I demoed this feature to him. I plan to implement that with a separate PR as I want to make this feature work first, and don't want this PR to grow any further. :)

pivotal-jwinters replied:
@evanchaoli is there a strong use case for wanting to cycle through all cred managers? It seems like we could get into some weird behaviour if someone added a new cred manager which unknowingly overrides a value from another one. And they wouldn't know it until it was already running in the pipeline.

I know it adds more to this PR, but I would be in favour of referencing cred managers by name as being the only supported method. It's much more in line with how resources work, and the direction we want to take with re-usable "resources/tasks" in `projects`.

evanchaoli replied:
@pivotal-jwinters I don't have a strong use case right now. One use case I can image is, it would be helpful for pipeline debugging. Like when debugging the pipeline, user may add a `dummy` var source in the first place, which would make sure all vars can be interpolated.

The other point is that, I don't think much user would add more than one var_source to a pipeline. Practically, maintaining a Vault is not easy given the serious security matter. So I don't think cycle through all var sources would bring serious issue. Only when users start to use the feature, then we know if that's really an issue.

To support `((myvault:foo/bar.baz))` syntax, some more changes are needed, because with current code, var_source name is not tracked by `creds.Secrets`. Again, I don't want to make this PR any bigger, it has been functional, and should fit most of use cases.

As our communication has long delay due to time difference, I just thought over, if you strongly insist to support var source name, a quick trade-off I can do is, to restrict only the first var_source defined in pipeline will be used. This trade-off would 1) eliminate the "cycle through" concern, 2) not impact our use case because our use will mostly add only one var_source to a pipeline, 3) still support the debugging use case as I described above.

To summarize, I still suggest merge this PR first. If you are strongly against that, I can do the trade-off. Once this PR is merged, I commit to contribute a PR to support `((myvault:foo/bar.baz))` syntax.

pivotal-jwinters replied:
@evanchaoli I understand that it's functional and that it meets all your use cases, however if we introduce named credential managers later on, it will be a breaking change. However since since this is going to be an experimental/undocumented feature maybe that's ok. I think as long as we get to the point where named credential managers are the only supported method before we remove the experimental tag, then I would be ok merging this as is.

vito replied:
👍 to merging for now and requiring named var references later - I'm working that aspect into my RFC. 🙂

evanchaoli replied:
@pivotal-jwinters Thanks for understanding. I'm absolutely fine with only removing the experimental tag after named syntax is supported. :) But to clarify, the next PR, for named var source, would NOT be a breaking change. The behavior I'm planning to implement is like:

* `((var1))`: iterate over all var_sources including the system level credential manager to look for `var1`
* `((name1:var1))`:  only look up var source `name1` for `var1`.
* `((name1,[name2,...]:var1))`: only look up the multiple specified var sources for `var1`.
* `((sys:var1))`: only look up the system level credential manager for `var1`
* `((name1,[name2,...,]sys:var1))`: only look up specified var sources as well as the system level credential manager for `var1`.

which would leave the flexibility to users. 

Anyway, please merge this PR first. If you have different opinion with the design, we may discuss in a new issue. Once this PR is merged, I'll create an issue to track the named var source change, and we may discuss the design over there.

vito commented:
Actually now I'm wondering if we should go through the RFC process for this. :thinking: There was a lot of nuance captured in https://github.com/concourse/rfcs/pull/21 and its comments, and it's adding pretty significant functionality to pipelines, which is typically what the RFC process is for.

I'm already going to have to wrap up all of our RFCs for v10 now that you've lit a fire under my ass with https://github.com/concourse/concourse/pull/4708 lol. I can write up a new RFC based on the team credential managers RFC, and have it reflect my understanding of this PR and maybe bring up some of the concerns/ideas from concourse/rfcs#21 so we can agree on the next steps to take after this is merged.

I don't want your branch to set around forever and go stale in the meantime so I'm OK with merging this anyway and having it exist as an experimental or undocumented feature until the RFC process completes, as long as we're sure it won't cause any issues.

evanchaoli commented on atc/api/resourceserver/check_webhook.go:
```diff
@@ -39,7 +39,13 @@ func (s *Server) CheckResourceWebHook(dbPipeline db.Pipeline) http.Handler {
 		}
 
-		variables := creds.NewVariables(s.secretManager, dbPipeline.TeamName(), dbPipeline.Name())
+		globalVariables := creds.NewVariables(s.secretManager, dbPipeline.TeamName(), dbPipeline.Name(), false)
```
Good point!

evanchaoli commented on atc/creds/pool_test.go:
```diff
@@ -0,0 +1,190 @@
+			Expect(err).ToNot(HaveOccurred())
+			Expect(varSourcePool.Size()).To(Equal(1))
+
+			time.Sleep(2*time.Second)
```
This is nice!

evanchaoli commented:
> I don't want your branch to set around forever and go stale in the meantime so I'm OK with merging this anyway and having it exist as an experimental or undocumented feature until the RFC process completes, as long as we're sure it won't cause any issues.

Thanks for understanding. I'm absolutely fine to have it as an experimental or undocumented feature for a while.

evanchaoli commented:
> @evanchaoli I'll have an RFC up soon that covers `((var))` semantics with regard to `var_sources` - so I would recommend holding off on further work until we can align on how that should behave. 🙂 Feel free to start anyway, but e.g. I'm currently thinking we should just enforce named var usage for `var_sources` and have `((foo))` _only_ query the global credential manager and support `((name:foo))` for `var_sources`. We can discuss this further in the RFC once I have a draft up.
> 

@vito Sure, I'll only start next step implementation after we align on the design. Do we need to open an issue and discuss design in the issue?

BTW, I just tried, unfortunately, I'm not able to move cards on project.

vito commented:
> @vito Sure, I'll only start next step implementation after we align on the design. Do we need to open an issue and discuss design in the issue?

I'm going to either submit a new RFC or adapt concourse/rfcs#21 to it. Going to try and get something up today. I'll comment here when it's up!

vito commented:
@evanchaoli submitted a draft here: https://github.com/concourse/rfcs/pull/39
